### PR TITLE
[mlir][ods] Verify access to operands in inferReturnTypes

### PR DESCRIPTION
**Pull request by npanchen**
The patch adds graceful handling of incorrectly constructed MLIR operation with less operands than expected.

### CODE REVIEW

File: mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp [hunk 1]
```diff
@@ -3583,6 +3583,22 @@ void OpEmitter::genTypeInterfaceMethods() {
   fctx.addSubst("_ctxt", "context");
   body << "  ::mlir::Builder odsBuilder(context);\n";
 
+  // Preprocessing stage to verify all accesses to operands are valid.
+  int maxAccessedIndex = -1;
+  for (int i = 0, e = op.getNumResults(); i != e; ++i) {
+    const InferredResultType &infer = op.getInferredResultType(i);
+    if (!infer.isArg())
+      continue;
+    auto arg = op.getArgToOperandOrAttribute(infer.getIndex());
```
Comment (weiweichen) [inline mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:3592]: spell `auto` here please

File: mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp [hunk 2]
```diff
@@ -3583,6 +3583,22 @@ void OpEmitter::genTypeInterfaceMethods() {
...
+    if (arg.kind() == Operator::OperandOrAttribute::Kind::Operand)
+      maxAccessedIndex =
```
Comment (weiweichen) [inline mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:3594]: [nit] add `{}` here since formatting has two lines

**Review by weiweichen — COMMENTED** [review]
LGTM with minor comments. Not sure if I can approve MLIR changes though.

**Review by joker-eph — APPROVED** [review]

File: mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp (hunk 1 above)
Comment (npanchen) [reply mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:3592]: Apparently, even `getArgToOperandOrAttribute ` is declared with `auto`, so no one wants to expose return type.

Comment (joker-eph) [reply mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:3592]: I see it declared as `  OperandOrAttribute getArgToOperandOrAttribute(int index) const;` ?

Comment (npanchen) [reply mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:3592]: my bad. I thought it's pure automatic type deduction, so didn't look for `->`

**Pull request merged**